10-applicative/AParser.hs

```diff
@@ -0,0 +1,248 @@
+• p1 <*> p2 represents the parser which first runs p1 (which will
+consume some input and produce a function), then passes the
+remaining input to p2 (which consumes more input and produces
+some value), then returns the result of applying the function to the
+value. However, if either p1 or p2 fails then the whole thing should
+also fail (put another way, p1 <*> p2 only succeeds if both p1 and
+p2 succeed).
+So what is this good for? Recalling the Employee example from
+class, type Name = String
+data Employee = Emp { name :: Name, phone :: String }
+we could now use the Applicative instance for Parser to make an
+employee parser from name and phone parsers. That is, if
+parseName :: Parser Name
+parsePhone :: Parser String
+then
+Emp <$> parseName <*> parsePhone :: Parser Employee
+is a parser which first reads a name from the input, then a phone
+number, and returns them combined into an Employee record. Of
+course, this assumes that the name and phone number are right
+next to each other in the input, with no intervening separators. We’ll
+see later how to make parsers that can throw away extra stuff that
+doesn’t directly correspond to information you want to parse
+-}
+--Exercise 2
+instance Applicative Parser where
+    pure a    = Parser $ const $ Just (a, [])
+    --Working for understanding:
+    --p1's runParser :: String -> Maybe((a -> b), String)
+    --p2's runParser :: String -> Maybe(a, String)
+    p1 <*> p2 = Parser $ \str -> case runParser p1 str of
+        Nothing           -> Nothing
+        Just (func, rest) -> runParser (fmap func p2) rest
+
+{-
+Exercise 3
+We can also test your Applicative instance using other simple
+applications of functions to multiple parsers. You should implement
+each of the following exercises using the Applicative interface to put
+together simpler parsers into more complex ones. Do not implement
+them using the low-level definition of a Parser! In other words, pretend that you do not have access to the Parser constructor or even
+know how the Parser type is defined.
+• Create a parser
+abParser :: Parser (Char, Char)
+which expects to see the characters ’a’ and ’b’ and returns them
+as a pair. That is,
+*AParser> runParser abParser "abcdef"
+Just ((’a’,’b’),"cdef")
+*AParser> runParser abParser "aebcdf"
+Nothing
+• Now create a parser
+abParser_ :: Parser ()
+which acts in the same way as abParser but returns () instead of
+the characters ’a’ and ’b’
+*AParser> runParser abParser_ "abcdef"
+Just ((),"cdef")
+*AParser> runParser abParser_ "aebcdf"
+Nothing
+• Create a parser intPair which reads two integer values separated
+by a space and returns the integer values in a list. You should use
+the provided posInt to parse the integer values.
+*Parser> runParser intPair "12 34"
+Just ([12,34],"")
+-}
+--Exercise 3
+abParser :: Parser (Char, Char)
+abParser = (,) <$> char 'a' <*> char 'b'
+
+abParser_ :: Parser ()
+abParser_ = (\a b -> ()) <$> char 'a' <*> char 'b'
```

Comment:
You can reuse abParser and just `fmap` the result to `()`.
Also check out `$>`, `<$` and `void` from base which allow you to do the same more briefly.

Reply:
@jhrcek
For the version using `void`, I'm not really sure how to make use of `$>` and `<$` to get it working.
I have a version making use of `abParser` again and that works with `void`.
```haskell
-- With fmap
abParser_ = fmap (const ()) abParser
-- With void
abParser_ = void abParser
```

Reply:
I should have made it clearer that shouldn't try to be using the suggested operators all at once.
Just to realize that there is an equivalence between:
```Haskell
fmap (\_ -> ()) x ==
fmap (const ()) x ==
() <$ x ==
x $> () ==
void x
```
All of these expressions are equivalent. They essentially "replace all occurrences of `a` withing `f a` with `()`".

Reply:
@jhrcek No worries, got it. Thank you! 👍